### ISSUE TITLE
objects in rhai

### ISSUE THREAD

**Issue by brianwp3000** [jonathandturner/rhai#73]
I see documentation on how to define structs in Rust and then make them available for use in Rhai. These structs can have fields that are accessible with dot notation, e.g:
```
let foo = new_ts(); // new_ts() is a Rust-defined fn that returns a Rust-defined struct
foo.bar = 5;
```
Is there any way to create an object in Rhai without first having to define it in Rust? I'm looking for something like Javascript's objects which allow you do something like this:
```
var foo = { bar: 0 };
foo.bar = 5;
```

Is this on the roadmap for Rhai?

**Comment by luciusmagn** [comment]
It's on the roadmap yes, but not possible yet, sorry

čt 21. 6. 2018 v 2:41 odesílatel Brian Porter <notifications@github.com>
napsal:

**Comment by zutils** [comment]
I would also like objects in Rhai.  I REALLY would like to create structure types in Rhai and analyze them in Rust (viewing the properties/name/types etc...)

**Comment by stevedonovan** [comment]
There is a fairly straightforward route here - first, provide support for maps as well as vectors. This seems to be easy enough, if the maps (HashMap) go from String to Any - then the indexing is unambiguous.  Second, allow 'object literals' like `{a = 1, c = "hello"}`.  The value of this is just `HashMap<String,Any>` which is easy enough to work with on the Rust side.

**Comment by stevedonovan** [comment]
(Although that's syntactically ambiguous - I'm thinking too much in Lua here. But `{a: 1}` could work)

**Comment by stevedonovan** [comment]
(Although that's syntactically ambiguous - I'm thinking too much in Lua here. But `{a: 1}` could work)

**Comment by stevedonovan** [comment]
Sorry, comment lost: it is syntactically possible to have Javascript-style object literals like `{one: 1, two: "hello"}` which map to a `HashMap<String,Box<Any>>`.  Can unambiguously extend `[]` to do lookup on such maps.

**Comment by jhwgh1968** [comment]
I've been playing around with Rhai for a project I'm doing, and found this as a roadblock. Just to try and understand the code, I started on a slightly different implementation:

```
let myarray = [0, 1, 2];
let mymap = [red: "value", orange: 4.0, yellow: 20];
```

Would this syntax be acceptable?

**Comment by schungx** [comment]
> Would this syntax be acceptable?

I'm quite sure an object syntax is possible, even not hard to implement.  But is it really necessary to burden a _scripting_ language with objects?

In Rhai, you can already register your own object types.  That means all objects you want to create can be created first in Rust, then registered for use in Rhai.

So in your example:

```rust
struct MyMap {
    pub red: String,
    pub orange: f64,
    pub yellow: i64
}

engine.register_type::<MyMap>();
engine.register_fn("new_map", || MyMap { red: "value", orange: 4.0, yellow: 20 });
engine.register_get_set("red", |x: &mut MyMap| x.red, |x: &mut MyMap, val: String| x.red = val);
engine.register_get_set("orange", |x: &mut MyMap| x.orange, |x: &mut MyMap, val: f64| x.orange = val);
engine.register_get_set("yellow", |x: &mut MyMap| x.yellow, |x: &mut MyMap, val: i64| x.yellow = val);
let mut scope = Scope::new();
scope.push_constant("mymap", );
```

In Rhai, you can then do things like:

```rust
let mymap = new_map();
let r = mymap.red;
mymap.yellow = mymap.orange + r.len().to_float();
print(mymap.yellow);
```

**Comment by schungx** [comment]
> let mymap = [red: "value", orange: 4.0, yellow: 20];

Since Rhai is a recursive-descent parser with one look-ahead, it will be a bit tricky to decide whether the expression is an array or an object.  Same with `{` ... `}` syntax, which cannot distinguish between a block statement or an object.

That's because the colon is two tokens after the wrapping token...

**Comment by schungx** [comment]
A simple implementation is probably to add support for `HashMap<String, Box<dyn Any>>` as a primitive type just like `Array` (which is `Vec<Box<dyn Any>>`).

Object literals can be parsed similar to array literal, and accessed via the dot or bracket notation.

Rust code accessing the object will be returned a `HashMap<String, Box<dyn Any>>`.

But my question remains: what potential use case will this solve that cannot be solved better by defining a Rust struct and registering it?  Other than being able to return a whole structure at once.

**Comment by profan** [comment]
Isn't the whole point of having a script language to be able to do work in the script language without having to recompile the source of what you're using it in?

In the case of not having any ability to create your own intermediate structures in script you'll end up having to recompile the whole system as soon as you realise you want even one struct more, that might just be used in a single place in the script, seems to defeat the purpose of script a little no?

**Comment by schungx** [comment]
That sounds reasonable.  I guess I'm not writing a game so I won't know the needs there!  I just need a scripting engine to write business rules...

The immediate problem is with the syntax.  The normal object literal syntax:

```rust
{ field1: value1, field2: value2, field3: value3 }
```

is LR(2), which may require a bit of work to splice into the current parser.  If I pick a different syntax, like using a symbol that's not used anywhere, I can make it 1-look-ahead, but then it'll be different from anything anybody is used to...

**Comment by profan** [comment]
Alternatively, maybe something like `#{}` for a hash literal? The "common" syntax would definitely be the most optimal, but I trust you'll come up with something reasonable if you don't want to turn the parser into a LR(2) one (which I reckon could actually be constructed as a LR(1) one still, but grammar probably being way more convoluted in that form).

**Comment by jhwgh1968** [comment]
> The normal object literal syntax... is LR(2), which may require a bit of work to splice into the current parser.

This is the reason I picked overloading the array syntax, @schungx.

My current prototype parses quite easily, since it is only LL(3).  The language grammar becomes (to simplify some edge cases):

```
expr =>   <rules for integers, floats, etc> |
               leftbracket (expr comma)* rightbracket  |
               leftbracket (ident colon expr comma)* rightbracket | 
               leftbracket colon rightbracket
```

The last rule handles empty maps, to distinguish them from empty arrays. (Even if array slicing is added, I presume we will not use the Python-like `[:]` to mean "copy the entire array".)

The key is that the look-ahead only needs to distinguish between three things after the bracket: `ident colon` (an illegal expression), `colon` (an illegal expression), or `expr`. I can do that just with a little extra logic in the expr parser, right before the `[` token calls `parse_array_literal`.

The main limitation of my version is that it currently only accepts string keys via identifiers. That fits my use case, which is to have user code be able to return JSON-like objects to the calling Rust code. I think it could be expanded, but I'm interested in getting something working first.

@profan, I also chose the `[]` syntax, because there are languages that also use it for maps, such as Perl (in certain circumstances) and PHP. The latter, in particular, overloads it because it uses an internal container which becomes an array when indexed with numbers, and a map when indexed with other types.

**Comment by jhwgh1968** [comment]
A second general note, @schungx.

It's not a game engine, but my use case boils down to the same motivation as many game engines have: user-defined _callback handlers_.

That is, the Rust code defines an API for some operation, and the scripting language allows users to write code to match it. In games, that often involves custom data structures going both ways.

While passing in "output objects" from Rust may work well enough, it denies the API designer a lot of flexibility, and makes the code harder to read. If you're not deeply familiar with the language, you can't inspect the code to tell what type something is, only do your own personal type inference analysis in your head and guess.

It becomes even more tricky if you want an API, for example, that could be a union of three different types, which come in under different circumstances or want different outputs.

In short, anonymous compound types (i.e. "objects" in the JSON and Javascript sense) are the easiest solution to this problem, and allow people to write better, simpler, cleaner code.

**Comment by jhwgh1968** [comment]
I've poked at this for a while, and found that making the `TokenIter` LL(n) -- where n = 1 most of the time, but n = 2 for a small piece of code -- is trickier than I thought.

I expected it to be easy by using [`itertools::multipeek`](https://docs.rs/itertools/0.9.0/itertools/fn.multipeek.html), but finding all the points I would need to `reset_peek` in the existing parser code turned out to be quite a challenge.

As a result, I have tweaked the syntax a little more to make it truly LL(1):
```
expr =>   <rules for integers, floats, arrays, etc> |
          leftbracket colon (ident colon expr comma)* rightbracket | 
```

Which means it now looks like this:
```
let myarray = [1, 2, 3];
let mymap = [: red: 1, orange: 2.0, yellow: "3"];
```

Any opinion on that compared to your suggestion of `#{}`, @profan?

**Comment by schungx** [comment]
It looks like you're parsing the first array/map identifier separately to turn it into LL(1).

Using this technique, it should be possible to use `{` ... `}` pairs for a more _normal_ looking object literal also, although the cases to consider in case of `{` will be many more than an array literal...  By careful surgery, it might be possible to override parsing of the unary expression term to detect a follow-on colon...

Alternatively, if we keep brackets, then maybe we consider using `[ prop => value ]` syntax similar to PHP?

But still, if we can get `{` ... `}` working at all, it might be best to keep the uniform feel because Rhai currently takes after Rust and JS...

**Comment by schungx** [comment]
Technically speaking, we can pick any symbol (used or not) that can never start a sub-expression (such as `<`).  Syntax like:

```<a=1, b=2>```

looks almost XML.  However, the matching ending token `>` _does_ form part of an expression, so

```< a=1, b=2 > y```

will be ambiguous, if we ever take away the requirement of separating statements with `;` (which it was originally optional).  I added the rule just because the semicolon is so easy to forget and it sometimes generate obscure bugs that are very hard to track down.  But if we want to relax this requirement in the future, this will generate a problem.

So I still think sticking with an _unused_ symbol is probably best.

**Comment by schungx** [comment]
> Rhai does not seem to support option types natively

We don't need objects for this.  All we need is a special API `engine.eval_to_optional::<T> -> Option<T>` that maps a return value in `T` to `Some<T>` and `()` to `None`.  In your script, you can then `throw ();` to exit with `None`.

**Comment by schungx** [comment]
I've put in a full implementation for object maps in my fork under the object_maps branch:

https://github.com/schungx/rhai/tree/object_maps

This branch:

* adds object map literals support under the `${ prop: value }` syntax
* supports dot-property access: `obj.prop = value; let x = obj.prop;`
* supports index access: `obj["prop"] = value; let x = obj["prop"];`  (similar to JS)
* some utility functions: `obj.prop = value; if obj.has("prop") { print(obj.prop); }`
* Reading a non-existent property returns `()` instead of causing an error

**Comment by jhwgh1968** [comment]
I'll try your branch with my use case, @schungx! I bet it will work just fine. The addition of `has` is also nice!

**Comment by schungx** [comment]
> The addition of `has` is also nice!

Lifted it almost straight from JS's `hasOwnProperty` LOL!

**Comment by schungx** [comment]
This PR https://github.com/jonathandturner/rhai/pull/120 adds object maps.

**Comment by profan** [comment]
I'd go for 3. personally, since you said $ is (relatively) commonly used in string interpolation type contexts, plus `prop: value` would fit in nicely with if one might want named arguments later and that could likely look similar ie. `func(arg1: v, arg2: v)` and would fit right in with it.

Then again you could make the same argument about `=` assuming you used the same thing in both places at least, but .. :thinking:

**Comment by schungx** [comment]
@profan   you have a good point about code reuse, but the relevant code sections are so short (it really isn't very hard to parse a literal) that it might not save much.

So why don't I change it to `#{ prop: value }` and update the PR.  Let's see how it goes (to paraphrase POTUS)...

**Comment by jhwgh1968** [comment]
Option 3 sounds okay to me!

I like the colon syntax better than the others, and you could say it's a "hash map" -- since "#" is sometimes called "hash" in UNIX land. :grin:

**Comment by schungx** [comment]
Since this is really not an _object_ as per OOP style (i.e. no encapsulation, inheritance and polymorphism), maybe I should change it to "hash map" instead of "object map" in the docs?

**Comment by schungx** [comment]
Off-topic: I have created a new issue to track a new feature: String interpolation

https://github.com/jonathandturner/rhai/issues/121

**Comment by schungx** [comment]
I have a minor commit to my fork that adds a very useful function: `mixin` to object maps.

Now you can do:

```
let x = ${ ... };
let y = ${ ... };
x.mixin(y);
x += y;
let z = x + y;
```

Lifted this directly from JS.
